### PR TITLE
Add reasons to on_dieplayer and on_hpchange

### LINKED ISSUE #1342: on_dieplayer and on_hpchange reasons
Wouldn't it be nice to have a reason given each time a player is damaged, e.g.
```
player:set_hp(damage, reason)

player:punch(puncher, time_from_last_punch, tool_capabilities, direction, reason)

minetest.register_on_player_hpchange(function(player, hp_change, reason)
```
... where 'reason' is a simple text string containing something like "lava", "fire", "mob_sandmonster", "ethereal:crystal_spike" etc.

This would come in very handy for award or death messages or even stats on a players game :)

-------------------------------

Instead of just a player the on_dieplayer should also pass the node or entitiy that killed it as a parameter.

A suggestion:

```
minetest.register_on_dieplayer(params                                                                              
  print(params.player) --the player that got killed
  print(params.killer) --returns the player that killed him/her
  print(params.node) --retruns the node that inflicted the final blow
end)
```

This might require for the set_hp(hp) function to also ask for extra information on the reason for the hp reduction.

### PR DESCRIPTION
**Pull request by rubenwardy**
```lua
minetest.register_chatcommand("a", {
	func = function(name)
		minetest.get_player_by_name(name):set_hp(10, {
			type = "fall",
			a = function()
				minetest.chat_send_all("THIS WORKS!")
			end
		})
	end
})

minetest.register_on_dieplayer(function(player, reason)
	minetest.chat_send_all("DIE " .. reason.type)
	print("DIE " .. player:get_player_name())
	print("DIE " .. dump(reason))
	if reason.a then
		reason.a()
	end
end)

minetest.register_on_player_hpchange(function(player, hp, reason)
	minetest.chat_send_all("HP " .. reason.type)
	print("HP " .. player:get_player_name())
	print("HP " .. dump(reason))
	if reason.a then
		reason.a()
	end
end)
```
--------------

> * `reason`: a PlayerHPChangeReason table.
>     * The `type` field will have one of the following values:
>         * `set_hp` - A mod or the engine called `set_hp` without
>                      giving a type - use this for custom damage types.
>         * `punch` - Was punched. `reason.object` will hold the puncher, or nil if none.
>         * `fall`
>         * `node_damage` - damage_per_second from a neighbouring node.
>         * `drown`
>         * `respawn`
>     * Any of the above types may have additional fields from mods.
>     * `reason.from` will be `mod` or `engine`.

-------------------------
Fixes #1342

To do:

- [x] `fall` type is probably wrong, it's just client declared damage. Need to see if there's any way to get this.
- [x] Add puncher to the punch reason
- [x] Investigate entity caused punches
- [x] Allow mods to specify a reason when doing `set_hp`
- [ ] Throw error on bad type in player:set_hp()

### CODE REVIEW

File: src/content_sao.cpp [hunk 1]
```diff
@@ -931,8 +931,9 @@ void PlayerSAO::step(float dtime, bool send_recommended)
 
 			// No more breath, damage player
 			if (m_breath == 0) {
-				setHP(m_hp - c.drowning);
-				m_env->getGameDef()->SendPlayerHPOrDie(this);
+				auto reason = PlayerHPChangeReason(PlayerHPChangeReason::DROWNING);
```
Comment (nerzhul) [inline src/content_sao.cpp]: wrong auto + affectation, poor performance. PlayerHPChangeReason reason(value)

File: src/content_sao.h [hunk 2]
```diff
@@ -413,3 +413,22 @@ class PlayerSAO : public UnitSAO
 	bool m_physics_override_new_move = true;
 	bool m_physics_override_sent = false;
 };
+
+
+struct PlayerHPChangeReason {
+	enum TYPE {
```
Comment (nerzhul) [inline src/content_sao.h]: don't call enum in upper case + add u8 size to enum (enum Type : u8 { )

File: src/content_sao.cpp (hunk 1 above)
Comment (rubenwardy) [reply src/content_sao.cpp]: oops, been coding Kotlin all day :D

File: src/network/serverpackethandler.cpp [hunk 3]
```diff
@@ -1478,12 +1479,12 @@ void Server::handleCommand_Interact(NetworkPacket* pkt)
 			// If the object is a player and its HP changed
 			if (src_original_hp != pointed_object->getHP() &&
 					pointed_object->getType() == ACTIVEOBJECT_TYPE_PLAYER) {
-				SendPlayerHPOrDie((PlayerSAO *)pointed_object);
+				SendPlayerHPOrDie((PlayerSAO *)pointed_object, PlayerHPChangeReason(PlayerHPChangeReason::PLAYER_PUNCH, playersao));
```
Comment (nerzhul) [inline src/network/serverpackethandler.cpp]: if i'm correct C++11 notation {PlayerHPChangeReason::PLAYER_PUNCH, playersao} can be used instead

File: src/network/serverpackethandler.cpp [hunk 4]
```diff
@@ -1478,12 +1479,12 @@ void Server::handleCommand_Interact(NetworkPacket* pkt)
...
 			}
 
 			// If the puncher is a player and its HP changed
 			if (dst_origin_hp != playersao->getHP())
-				SendPlayerHPOrDie(playersao);
+				SendPlayerHPOrDie(playersao, PlayerHPChangeReason(PlayerHPChangeReason::PLAYER_PUNCH, pointed_object));
```
Comment (SmallJoker) [inline src/network/serverpackethandler.cpp]: These two lines are way too long.

Comment (rubenwardy) [reply src/network/serverpackethandler.cpp]: I'd prefer implementation comments rather than long lines or bad braces :P

**Comment by hkzorman** [comment]
This is great! Thanks for taking time to do this.
Some observations:
* `punch` shouldn't be limited to player. It could be an entity hitting the player. 
  * In addition, I believe #1342 actually asks for providing a custom reason when calling `punch`
* Can it be possible to add a custom reason type in addition to the defined ones?

**Comment by rubenwardy** [comment]
> punch shouldn't be limited to player. It could be an entity hitting the player. 

Oh, I didn't realise this was possible. I'll change the docs, it's probably already possible.

> Can it be possible to add a custom reason type in addition to the defined ones?

I'm planning a custom type where mods can provide a table like so:

```lua
{
    type = "custom",
    value = {
        foo = "bar"
    }
}
```

I'm not sure about adding it to every one a mod can cause. I'll have to think about that.

**Comment by hkzorman** [comment]
Now that you mention it, I might be mistaken. I know anything can call
player:punch(), so maybe the best thing to do is to actually add another
parameter to player:punch() so that whatever calls it, provides the reason?

Adding the custom type with a table is a good idea. This allows maximum
flexibility for mods.

**Comment by raymoo** [comment]
In my mind, `punch` just means you want to damage an SAO using the damage-armor system, with no actual implication of a physical puncher. From this perspective it makes sense to be able to provide a custom reason.

File: src/network/serverpackethandler.cpp (hunk 3 above)
Comment (rubenwardy) [reply src/network/serverpackethandler.cpp]: this doesn't scale when adding more parameters (which I may do)

**Comment by rubenwardy** [comment]
Now I just need to let mods provide reasons.

I suggest adding a popPlayerHPChangeReason which gets the reason from the function call. So you could do this:

```lua
player:set_hp(player:get_max_hp(), {
    type = "respawned"
})

player:set_hp(player:get_hp() - 3, {
    type = "custom",
    foo = "bar",
    bob = {
        dillan = true
    }
})
```

* Should mods be able to add custom fields to all reasons?
* What stops a mod adding a field which is later used by core?
* My plan is to JSON encode custom fields, but this would stop user data and functions being passed.

**Comment by raymoo** [comment]
@rubenwardy 
>Should mods be able to add custom fields to all reasons?

Probably only the reasons that are expected to be mod-caused.

>What stops a mod adding a field which is later used by core?

The API could strongly recommend a "modname:typename" naming convention. It would be difficult to enforce since the mod name might not be known when the reasons is made (unless reason types have to be preregistered), but I guess you could cheat and define not following the convention as UB.

>My plan is to JSON encode custom fields, but this would stop user data and functions being passed.

What's the purpose of JSON encoding?

**Comment by rubenwardy** [comment]
> What's the purpose of JSON encoding?

To allow tables in the data structure, and to add types

**Comment by raymoo** [comment]
Are the reasons being sent over the network that they can't just be Lua values?

**Comment by nerzhul** [comment]
@raymoo  sending them over network permit to catch them in CSM mods

**Comment by rubenwardy** [comment]
> Are the reasons being sent over the network that they can't just be Lua values?

It goes through the c API, so you can't store Lua values - unless you hackily store it in a temporary value

**Comment by raymoo** [comment]
@rubenwardy Doesn't the Lua API provide `luaL_ref` (and the regisry more generally) for this purpose?

**Comment by rubenwardy** [comment]
@raymoo this now uses luaL_ref, thanks for the tip

**Comment by nerzhul** [comment]
@rubenwardy i suggest to add an integration test in experimental for this, permitting to use your exampls and test the features

**Comment by nerzhul** [comment]
Okay for me

**Pull request merged**